Sources/Datadog/RUM/RUMVitals/VitalRefreshRateReader.swift

```diff
@@ -0,0 +1,94 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import UIKit
+
+/// A class reading the refresh rate (frames per second) of the main screen
+internal class VitalRefreshRateReader {
+    private var observers = [VitalObserver]()
+    private var displayLink: CADisplayLink?
+    private var lastFrameTimestamp: CFTimeInterval?
+
+    init(notificationCenter: NotificationCenter = .default) {
+        notificationCenter.addObserver(
+            self,
+            selector: #selector(appWillResignActive),
+            name: UIApplication.willResignActiveNotification,
+            object: nil
+        )
+        notificationCenter.addObserver(
+            self,
+            selector: #selector(appDidBecomeActive),
+            name: UIApplication.didBecomeActiveNotification,
+            object: nil
+        )
+
+        start()
+    }
+
+    deinit {
+        stop()
+    }
+
+    /// `VitalRefreshRateReader` keeps pushing data to its `observers` at every new frame.
+    /// - Parameter observer: receiver of refresh rate per frame.
+    func register(_ observer: VitalObserver) {
+        DispatchQueue.main.async {
+            self.observers.append(observer)
+        }
+    }
+
+    /// `VitalRefreshRateReader` stops pushing data to `observer` once unregistered.
+    /// - Parameter observer: already added observer; otherwise nothing happens.
+    func unregister(_ observer: VitalObserver) {
+        DispatchQueue.main.async {
+            self.observers.removeAll { existingObserver in
+                return existingObserver === observer
+            }
+        }
+    }
+
+    // MARK: - Private
+
+    @objc
+    private func displayTick(link: CADisplayLink) {
+        if let lastTimestamp = self.lastFrameTimestamp {
+            let frameDuration = link.timestamp - lastTimestamp
+            let currentFPS = 1.0 / frameDuration
+            // NOTE: RUMM-1278 `oldValue` is not used
+            observers.forEach {
+                $0.onValueChanged(oldValue: 0.0, newValue: currentFPS)
+            }
+        }
+        lastFrameTimestamp = link.timestamp
+    }
```

Comment:
[We should avoid doing anything on the main thread](https://github.com/DataDog/dd-sdk-ios/blob/master/ZEN.md#small-footprint). Here, I can see two main thread penalties:
* 1/ In managing the array of observers (`observers.append()` and `observers.removeAll()`) on the `main` queue.
* 2/ In calling `onValueChanged()` on each observer, which runs [this `VitalInfo` computation](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/RUM/RUMVitals/VitalObserver.swift#L25-L42) and [notifies the `VitalListener`](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/RUM/RUMVitals/VitalObserver.swift#L41).

I assume that the `RUMViewScope` will be implementing the `VitalListener` interface, so it will contain:
```swift
func onVitalInfo(info: VitalInfo) {
   // ...
}
```
which will be run on the main thread. The rest of logic in `RUMViewScope` is executed [on the `background` queue](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/RUMMonitor.swift#L130-L134). 

I don't see any clear link nor plan on how this could be synchronized later without refactoring this `VitalRefreshRateReader`. The `info` would have to be accessed synchronously [when creating the `RUMViewEvent`](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/RUM/RUMMonitor/Scopes/RUMViewScope.swift#L304), but we can't perform `DispatchQueue.main.sync {}` as it will produce a hitch for each view update.

What is the idea for plugging this reader into `RUMViewScope`? 

IMO we should change the synchronisation in this class and dispatch things to background thread as early as possible to simplify further considerations and make it "thread safe" for the rest of architecture. These problems are already solved in [`NetworkInfoProvider`](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/Core/System/NetworkConnectionInfoProvider.swift#L61-L93), [`CarrierInfoProvider`](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/Core/System/CarrierInfoProvider.swift#L74-L110) and [`UserInfoProvider`](https://github.com/DataDog/dd-sdk-ios/blob/01e4c856971a8e530b1cd3d823e467895e9544c6/Sources/Datadog/Core/Attributes/UserInfoProvider.swift#L13-L36) with the Publish / Subscribe model and utilities introduced in https://github.com/DataDog/dd-sdk-ios/pull/422. Could we reuse the same concepts?